hyperapp.d.ts

```diff
@@ -0,0 +1,149 @@
+ * @param attributes     Any valid HTML atributes, events, styles, and meta data
+ * @param children  The children of the VNode
+ * @returns A VNode tree.
+ *
+ * @memberOf [VDOM]
+ */
+export function h<Attributes>(
+    nodeName: Component<Attributes> | string,
+    attributes?: Attributes,
+    ...children: (Children | Children[])[]
+): VNode<Attributes>
+
+/** @namespace [App] */
+
+/** An effect as the result of an ation
+ * 
+ * @memberOf [App]
+ */
+export type Dispatch<State> = <Data = {}>(obj: Action<State, Data> | State, data: Data) => void;
+
+
+/** An effect as the result of an ation
+ * 
+ * @memberOf [App]
+ */
+export type EffectFunc<State, Props, Payload, Result = void> = (dispatch: Dispatch<State>, props: Props, payload: Payload) => Result;
+
+/** An effect as the result of an ation
+ * 
+ * @memberOf [App]
+ */
+export type Effect<State, Payload> = [EffectFunc<State, unknown, Payload>, unknown]; // TODO: Use existential type props
+
+/** The result of an action.
+ *
+ * @memberOf [App]
+ */
+export type ActionResult<State, Payload> = State | [State, ...Effect<State, Payload>[]];
+
+/** The interface for a single action implementation.
+ *
+ * @memberOf [App]
+ */
+export type ActionFunc<State, Payload> = (state: State, data: Payload) => ActionResult<State, Payload>
+
+/** A reference to an action to be invoked by Hyperapp, with optional additional parameters
+ * 
+ * @memberOf [App]
+ */
+export type Action<State, Data = void> =
+    | [ActionFunc<State, unknown>, (unknown | ((data: Data) => unknown))] // TODO: Proper handling w/ existential types!
+    | ActionFunc<State, Data>;
+
+/** A reference to an subscription to be managed by Hyperapp, with optional additional parameters
+ * 
+ * @memberOf [App]
+ */
+export type Subscription<State> = [EffectFunc<State, unknown, void, () => void>, unknown]; // TODO: Existential types!
+
+/**
+ * A function used to create a particular subscription.
+ * 
+ * @memberOf [App]
+ */
+export type SubscriptionConstructor<State, Props = {}> = (props: Props) => Subscription<State>
+
+
+/** The view function describes the application UI as a tree of VNodes.
+ * @returns A VNode tree.
+ * @memberOf [App]
+ */
+export interface View<State> {
+    (state: State): VNode<object> | null;
+}
+
+/** The possible response types for the subscription callback for an application
+ * 
+ * @memberOf [App]
+ */
+export type SubscriptionsResult<State> = | (Subscription<State> | boolean)[] | Subscription<State>;
+
+/** The subscriptions function describes the current application subscriptions.
+ * @returns The current subscription(s) given the current state
+ * @memberOf [App]
+ */
+export type Subscriptions<State> = (state: State) => SubscriptionsResult<State>;
+
+/** The set of properties that define a Hyperapp application.
+ * @memberOf [App]
+ */
+export interface App<State> {
+    init?: (() => State) | State;
+    view: View<State>;
+    node?: Element;
+    subscriptions: Subscriptions<State>;
```

Comment:
`subscriptions` is optional. Also, as long as you read the hyperapp V2 code, `node` looks like it is required. So, isn't it correct to do this section as follows?

```ts
    node: Element;
    subscriptions?: Subscriptions<State>;
```

Reply:
Probably, yeah.